### PR TITLE
SBND PMT workflow

### PR DESCRIPTION
**Pull request by fjnicolas**
This PR includes the SBND workflow for the PMT.
- Main change:
  - Decoder for SBND and splitting the workflows betwen ICARUS/SBND (by @VCLanNguyen, details in [doc-30300])(https://sbn-docdb.fnal.gov/cgi-bin/sso/ShowDocument?docid=30300)
- Other updates
  - Added FFTs for the PMTs
  - Working fhicl configuration (test results in [doc-31775](https://sbn-docdb.fnal.gov/cgi-bin/sso/ShowDocument?docid=31775) slides 17-28)

### CODE REVIEW

File: sbndqm/dqmAnalysis/PMT/ICARUS/CAENV1730Streams_module.cc
Comment (VCLanNguyen) [inline sbndqm/dqmAnalysis/PMT/ICARUS/CAENV1730Streams_module.cc]: This needs to revert back "icarus-db01"

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (PetrilloAtWork) [inline sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: It looks like this is the only one you are using in the code:
```suggestion
#include "fhiclcpp/types/Sequence.h"
```

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (PetrilloAtWork) [inline sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: `art::Event::getByLabel()` is supposedly deprecated by _art_, the new pattern may be:
```suggestion
    art::Handle<artdaq::Fragments> thisHandle
      = event.getHandle<std::vector<artdaq::Fragment>>(input_tag);
    if( !thisHandle.isValid() || thisHandle->empty() ) continue;
```
The check of the first `if` in the original code was redundant with the second `if`.

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (PetrilloAtWork) [inline sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: The argument should be passed by reference to avoid a pointless and expensive copy; and if I am not wrong, it does not need to change the module object, so it should be declared `const`:
```suggestion
artdaq::Fragments daq::DaqDecoderSBNDPMT::readFragments( std::vector<art::Handle<artdaq::Fragments>> const& handles ) const {
```
(this change must be reflected also in the class definition above)

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (PetrilloAtWork) [inline sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: Suggestion: this code assumes that each handle has at least one fragment in it; this is guaranteed by the caller, so there is no actual problem, but to document that this is a requirement and to have a check on it we can add an assertion:
```suggestion
    assert(!handle->empty());
```
(it needs to `#include <cassert>`)

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (PetrilloAtWork) [inline sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: A detail: here there are two copies of each fragment. Two options to avoid them:
```suggestion
      for( auto const& frag : *handle ) {
	  
	       fragments.emplace_back( frag );
      }
```
or
```suggestion
      for( auto frag : *handle ) {
	  
	       fragments.emplace_back( std::move(frag) );
      }
```

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (PetrilloAtWork) [inline sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: This mask in ICARUS has hystorical reasons: PMT fragments had an ID like `0x0`, `0x1`, ... to `0x17`. Then with the integration of all subsytems we assigned an offset to the PMT fragments too, and they become `0x2000`, `0x2001` etc, but at that point ICARUS had some database and codebase with the old numbers; so we decided to remove the offset in our code.
Since you don't have that constrain, you may consider whether you want that or not.

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (PetrilloAtWork) [inline sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: My recurrent recommendation is that all the variables, including these local ones, which are not expected to be changed be declared `const`: it helps a lot to detect logic errors.

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (PetrilloAtWork) [inline sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: Note that ICARUS never disabled channels, so this management of disabled channels is tentative and not verified. If you plan to actually disable channels, you will want to check specifically this treatment.

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (PetrilloAtWork) [inline sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: I recommend no message be sent to `std::cout`.
In this case, the option `--trace` or an equivalent FHiCL configuration parameter may report the entry in each `produce()` module, including this one, so I would just remove it.

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (PetrilloAtWork) [inline sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: This PMT ID is almost certainly not going to stay the same as the LArSoft channel ID. I strongly recommend to insert a layer of mapping here, best if with a database. The database can be local (I think ICARUS stores it as SQLite file in `icarus_data`) to preempt network access to databases.

File: sbndqm/dqmAnalysis/PMT/ICARUS/fcl/pmt_online_monitor.fcl
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/ICARUS/fcl/pmt_online_monitor.fcl]: I am not sure who is the right contact person (@dtorretta56 or @mvicenzi?), but this FHiCL file should be renamed to include `icarus` in the name.

File: sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc]: Suggesting `constexpr` instead of `const`, and `static` too.
```suggestion
      static constexpr unsigned int nChannelsPerBoard = 16;
      static constexpr unsigned int nBoards = 24;
```

File: sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc]: The argument of this one should be a constant reference, and the function may be declared `static` since it does not depend on anything in the class:
```suggestion
        template<typename T> static T Median( std::vector<T> const& data );
```
That would also require changes in the implementation though.

File: sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc]: Contemporary _art_:
```suggestion
  art::Handle opdetHandle
   = evt.getHandle<std::vector<raw::OpDetWaveform>>( m_opdetwaveform_tag);
```

File: sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc]: Sugest to mark this as `override`:
```suggestion
        virtual void analyze(art::Event const & evt) override;
```

File: sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc]: There is quite some mix of naming standards here... if would be nice to stick to one.

File: sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc]: Here the exception-throwing access `.at()` is redundant since the control variable `i` is in the exact range of validity of the waveform vector (`NADC` was defined to be the waveform size), so an unchecked access is no worse and quite faster:
```suggestion
        *input = (double) opdetwaveform.Waveform()[i];
```

File: sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc]: Introducing `std::hypot()`:
```suggestion
        adcsFFT.push_back( std::hypot(real,im) );
```

File: sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc]: The logic here is not fully clear to me.
It looks like we are trying to get one waveform per channel (including the ones that might be disconnected/disabled)?
Do you really not care about which of the waveforms you pick within each channel?
Incidentally, here an implementation with a `std::vector<bool>` to track which channels have been seen and a counter tracking how many is quite more efficient than an unsorted vector like in this implementation.
Also, the prefix `m_` should be reserved for data members, which this variable is not.

File: sbndqm/dqmAnalysis/PMT/SBND/CMakeLists.txt
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/CMakeLists.txt]: I think  most of these Cmake variables are not set at all.
Judging from the module source code, a proper link list might look like:
```suggestion
  lardataobj::RawData
  larana::OpticalDetector_OpHitFinder
  sbndaq_online_storage
  sbndaq_online_redis_connection
  sbndaq_online_hiredis
  sbndqm::Decode_PMT_PMTDecodeData
  tpcAnalysis_SBN
```

File: sbndqm/dqmAnalysis/PMT/SBND/fcl/pmt_online_monitor_SBND.fcl
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/fcl/pmt_online_monitor_SBND.fcl]: I think `TFileService` is not used by any of the modules you are running.
```suggestion
```

File: sbndqm/dqmAnalysis/PMT/SBND/fcl/pmt_online_monitor_SBND.fcl
Comment (PetrilloAtWork) [inline sbndqm/dqmAnalysis/PMT/SBND/fcl/pmt_online_monitor_SBND.fcl]: I would omit this section entirely. It's ok for testing, but writing a "temporary" input file in a general configuration is not a great idea. In addition, if no `source` section is present _art_ will assume `RootInput` as source module every time an input file is specified, so there is not much difference anyway.
```suggestion
```

**Review by PetrilloAtWork — COMMENTED** [review]
I understand that the code in this pull request is heavily inspired by the existing ICARUS code, so most of my observations apply to the existing code as well. I have written many of those, because... well, because it's me. I invite you to consider these suggestions and make your own decision.
There is no major issue with the code that I could spot, and the only actual issue is something that should be solved by ICARUS (that is the FHiCL name collision).

I assume that this PMT decoder is going to be a simplified version for the data quality monitor rather than the offline one (ICARUS is trying to expose `icaruscode` to the DQM codebase so that we can use the official one, but that's not going great).
I do recommend that the channel ID produced here be the same as the off-line, LArSoft channel ID, to avoid endless confusion; in this code the channel mapping is hard-coded and I doubt that mapping can survive the challenge of the real detector (swapped cables, conflicting conventions... LArSoft has a very specific convention on channel numbering that you _should_, but you do not _have to_, follow).

In terms of formal approval, this is not one; go through the comments, drop a comment in the spectrum from "I like it as it is" to "I'll do how you suggest" and any form of discussion or clarification in between, and after that I'll add my formal approval.

**Comment by PetrilloAtWork** [comment]
Also, I apologise for the late review: it took me too long to get to it.

File: sbndqm/dqmAnalysis/PMT/ICARUS/fcl/pmt_online_monitor.fcl
Comment (mvicenzi) [reply sbndqm/dqmAnalysis/PMT/ICARUS/fcl/pmt_online_monitor.fcl]: I agree with the renaming, but I'm not expert enough to know where it is being referenced. I will need to investigate a little. More importantly, what happened to `pmt_online_monitor_configuration.fcl`? Has it been deleted? It seems its content is now at the top of `pmt_online_monitor.fcl`... why isn't github reporting that change?

Ohh.. I see. This branch is **behind** `develop`! `pmt_online_monitor_configuration.fcl` has not been added yet... This is extremely confusing. @PetrilloAtWork Maybe we can wait on the renaming until we merge? Many of the comments you have here apply also to the ICARUS side. We should probably have another PR after this one reviewing that as well...

Comment (PetrilloAtWork) [reply sbndqm/dqmAnalysis/PMT/ICARUS/fcl/pmt_online_monitor.fcl]: Ok for me, although I am not sure which is the (most) effective way to proceed.

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (fjnicolas) [reply sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: solved with 6b5231e

File: sbndqm/dqmAnalysis/PMT/ICARUS/CAENV1730Streams_module.cc
Comment (fjnicolas) [reply sbndqm/dqmAnalysis/PMT/ICARUS/CAENV1730Streams_module.cc]: see cbd9317

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (fjnicolas) [reply sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: removed remaining couts in 611aefc

File: sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc
Comment (fjnicolas) [reply sbndqm/dqmAnalysis/PMT/SBND/CAENV1730StreamsSBND_module.cc]: unify naming scheme in de46dba

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (fjnicolas) [reply sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: cbe1081

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (fjnicolas) [reply sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: f777f67

File: sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc
Comment (fjnicolas) [reply sbndqm/Decode/PMT/SBND/DaqDecoderSBNDPMT_module.cc]: see f777f67

File: sbndqm/DatabaseStorage/Connect.h
Comment (mvicenzi) [inline sbndqm/DatabaseStorage/Connect.h]: Is this `Connect.h` actually used? I can't find any active `#include path/to/Connect.h` anywhere...
There is no issue in changing these defaults, but I would like to make sure we are not relying on them somewhere else and we have the correct settings in the fcl file.

**Review by mvicenzi — APPROVED** [review]
I have reviewed the changes that relate to ICARUS, and I see no integration issues.
I just have one question that I'm adding below.

Several code suggestions by @PetrilloAtWork should also be implemented in the corresponding ICARUS flow, but it's better to do it AFTER this PR is merged.

**Comment by fjnicolas** [comment]
Thank you so much to @PetrilloAtWork for the extensive and detailed review! I implemented most of the code suggestions. The remaining ones (mainly related to the channel mapping and what waveforms to pick for each event) require further development and testing (but unfortunately, I'm no longer wroking on DQM neither I am in person at FNAL to test), so I'd say we should move forward with what we have for the shake os having a working configuration for the start of the SBND commissioning. I will open an issue with the "to do" list so people taking over with the PMT DQM is taskforce are aware.

**Comment by mvicenzi** [comment]
Most of this PR was merged in `release/v1_03_00` already with https://github.com/SBNSoftware/sbndqm/commit/9571e18ededf65ec66d5359e1fff5b6007d12a9d.
Merging what's left and closing.

**Pull request merged**
